src/main/kotlin/cacophony/utils/Diagnostics.kt

```diff
@@ -1,9 +1,16 @@
 package cacophony.utils
+import cacophony.token.Token
 
 interface Diagnostics {
     fun report(
         message: String,
         input: Input,
         location: Location,
     )
+
+    fun <TC : Enum<TC>> report(
```

Comment:
Nah I don't like this, I feel like `Diagnostics` should be generic, like a logger. Maybe the thing that is a specification is that it collects errors about `input`.

if you want to pass specific information in `report`, do it in `message`

instead of token, it could receive two locations that form a range